serving/src/main/java/feast/serving/service/OnlineServingService.java

```diff
@@ -64,110 +62,211 @@ public GetFeastServingInfoResponse getFeastServingInfo(
+        List<FeatureRow> featureRows = retriever.getOnlineFeatures(entityRows, featureSetRequest);
+        // Check that feature row returned corresponds to a given entity row.
+        if (featureRows.size() != entityRows.size()) {
+          throw Status.INTERNAL
+              .withDescription(
+                  "The no. of FeatureRow obtained from OnlineRetriever"
+                      + "does not match no. of entityRow passed.")
+              .asRuntimeException();
         }
+
+        for (var i = 0; i < entityRows.size(); i++) {
+          FeatureRow featureRow = featureRows.get(i);
+          EntityRow entityRow = entityRows.get(i);
+          // Unpack feature field values and merge into entityValueMap
+          boolean isStaleValues = this.isStale(featureSetRequest, entityRow, featureRow);
+          Map<String, Value> valueMap =
+              this.unpackValueMap(featureRow, featureSetRequest, isStaleValues);
+          entityValuesMap.get(entityRow).putAll(valueMap);
+
+          // Generate metadata for feature values and merge into entityFieldsMap
+          boolean isNotFound = featureRow == null;
+          Map<String, FieldStatus> statusMap =
+              this.getMetadataMap(valueMap, isNotFound, isStaleValues);
+          entityStatusesMap.get(entityRow).putAll(statusMap);
+
+          // Populate metrics
+          this.populateStaleKeyCountMetrics(statusMap, featureSetRequest);
+        }
+        this.populateRequestCountMetrics(featureSetRequest);
       }
 
-      List<FieldValues> fieldValues =
-          featureValuesMap.values().stream()
-              .map(valueMap -> FieldValues.newBuilder().putAllFields(valueMap).build())
+      // Build response field values from entityValuesMap and entityStatusesMap
+      // Reponse field values should be in the same order as the entityRows provided by the user.
+      List<FieldValues> fieldValuesList =
+          entityRows.stream()
+              .map(
+                  entityRow -> {
+                    return FieldValues.newBuilder()
+                        .putAllFields(entityValuesMap.get(entityRow))
+                        .putAllStatuses(entityStatusesMap.get(entityRow))
+                        .build();
+                  })
               .collect(Collectors.toList());
-      return getOnlineFeaturesResponseBuilder.addAllFieldValues(fieldValues).build();
+      return GetOnlineFeaturesResponse.newBuilder().addAllFieldValues(fieldValuesList).build();
     }
   }
 
-  private void populateStaleKeyCountMetrics(String project, FeatureReference ref) {
-    Metrics.staleKeyCount.labels(project, ref.getName()).inc();
-  }
+  /**
+   * Unpack feature values using data from the given feature row for features specified in the given
+   * feature set request.
+   *
+   * @param featureRow to unpack for feature values.
+   * @param featureSetRequest for which the feature row was retrieved.
+   * @param isStaleValues whether which the feature row contains stale values.
```

Comment:
Why do we pass this as an argument here? Seems like a code smell if the unpack needs to know something about staleness.